### PR TITLE
bugfix: persist accumulator data after reload_modules, fix #8881

### LINKED ISSUE #8881: file.managed & file.blockreplace using file.accumulated do not support reload
See the commit sls, problem is that when we issue a reload in between a file.accumulated state & its file.managed or file.blockreplace counterpart, the file.accumulated data structure is blanked.

This is due to the reload mecanism which reloads all the modules, then resetting any module variable setted in specific modules.

We should then either document not to use dynamic global module variables like in file, dockerio at least, or instruct a way to persist data in them or just have another clever way or reloading module, either by pickling global vars or having only explicit reload not to have bad surprises.

### PR DESCRIPTION
**Pull request by hvnsweeting**

### CODE REVIEW

File: salt/states/file.py [hunk 1]
```diff
@@ -244,6 +245,71 @@
...
+    return os.path.join(_get_accumulator_dir(), str(os.getpid()))
+
+
+def _cleanup_old_accummulator_data():
+    '''
+    Only keep current accumulator data.
+    '''
+    for fn in os.listdir(_get_accumulator_dir()):
+        path = os.path.join(_get_accumulator_dir(), fn)
+        if path != _get_accumulator_filepath():
```
Comment (kiorky) [inline salt/states/file.py]: test also that pid does not exist anymore, if you call a salt-call and a salt-master or something else linked to the same config exists on the same box, you can have a race condition where you would delete the other process accumulated data.

**Comment by thatch45** [comment]
@hvnsweeting I am still looking over this and wondering if there might be a better way

**Comment by hvnsweeting** [comment]
@kiorky  @thatch45  I changed the code to use id of State object instead of pid, please have a look. Thanks

**Comment by cachedout** [comment]
Hi @hvnsweeting 

@thatch45 asked me to take a look at this. I have a couple of requests:

1) I would like this feature to be config-gated and to default to False. If users wish to utilize this functionality, they should explicitly enable it.

2) Please do not use globals.

3) Please account for the running user when writing out the cache file so that we can account for cases where the minion might be running under different users (or with salt-call). This could be namespacing the files with the username or any other approach that's user-aware.

Let me know if you have questions. Otherwise, looks good! Thanks.

**Comment by hvnsweeting** [comment]
> 1) I would like this feature to be config-gated and to default to False. If users wish to utilize this functionality, they should explicitly enable it.

I don't agree with this. It is a bug - as why this issue (#8881) was created, and is a bug due to documentation: 

> Multiple accumulators can be "chained" together.
http://docs.saltstack.com/en/latest/ref/states/all/salt.states.file.html#salt.states.file.accumulated

So I don't think it need to configurable.

2) Please do not use globals.

I changed in the latest commit.

3) Please account for the running user when writing out the cache file so that we can account for cases where the minion might be running under different users (or with salt-call). This could be namespacing the files with the username or any other approach that's user-aware.

Not sure I understand what else need to be done, the cache file's name is ID of State object, which is unique between runs (state.sls/ state.highstate... ), then it cannot be duplicated as I see. Anything wrong?

**Comment by thatch45** [comment]
Sounds good @hvnsweeting, can you please make this commit to 2014.7? Otherwise it will not be merged forward

**Pull request merged**